<div style="border:solid Chocolate 2px; padding: 40px">

Привет Илья, меня зовут Саша Атаков. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех👍:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка❌:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<b><font color='orange'>Вот тебе полезная ссылка, вдруг подчерпнешь что-то новое для себя):</font></b>

* О логистической регрессии: https://habr.com/ru/company/io/blog/265007/
* Ну и случайный лес: https://alexanderdyakonov.wordpress.com/2016/11/14/%D1%81%D0%BB%D1%83%D1%87%D0%B0%D0%B9%D0%BD%D1%8B%D0%B9-%D0%BB%D0%B5%D1%81-random-forest/
* О dummy model - https://scikit-learn.org/stable/modules/generated/sklearn.dummy.DummyClassifier.html
* Более подробно о dummy model - https://www.geeksforgeeks.org/ml-dummy-classifiers-using-sklearn/
    
<b><font color='orange'>Общее впечатление</font></b>
* Большое спасибо за проделанную работу. Видно, что ты отлично освоил тему.
* Отлично, что ноутбук хорошо структурирован. Приятно проверять такие работы.
* В работе я оставил несколько советов. Буду рад, если ты учтешь их.
* В конце прописал свой комментарий по всему проекту, ознакомься с ним, пожалуйста)
* С радостью отвечу на твои вопросы, касательно выданной тебе темы, если они есть. Жду новую версию проекта.)

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 

Крайне полезно то, что ты написал описание к работе, потому что так можно точно понять, о чем идет речь в данном проекте - молодец)
</div>

## Откройте и изучите файл

In [1]:
# Импортируем необходимые библиотеки для работы (буду добавлять в эту ячейку нужные по мере необходимости)
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 

Крайне удобно, когда программист указывает все библиотеки с которыми он работал в самом начале проекта. Молодец!
</div>

In [2]:
# Читаем файл, смотрим первые 5 строк для изучения
df = pd.read_csv('/datasets/users_behavior.csv')
print(df.head(5))
print()
print(df.info())

   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
None


Данные успешно импортированы. У нас присутствует 3214 наблюдений (объектов) 5 различных переменных (признаков), из которых 4 переменные (признаки) имеют тип float, а 1 - тип int.

Среди данных мы имеем:
- сalls — количество звонков,
- minutes — суммарная длительность звонков в минутах,
- messages — количество sms-сообщений,
- mb_used — израсходованный интернет-трафик в Мб,
- is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 

Класс, хорошо расписал анализ того, какие данные у нас есть)
</div>

In [3]:
# Переведем данные о звонках и сообщениях к типу Int
df['calls']= df['calls'].astype('Int64')
df['messages']= df['messages'].astype('Int64')

Поскольку мы решаем задачу классификации, то мы можем рассмотреть следующие модели:
- Модель дерева решений
- Модель случайного леса
- Модель логистической регрессии

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 

Касательно изменения типа данных - абсолютно верно сделал, молодец). Про выбор моделей тоже все хорошо, ты постарался)
</div>

## Разбейте данные на выборки

Данные будут разбиты на выборки в следующем соотношении:
- Учебная - 60%
- Тестовая - 20%
- Валидационная - 20%

In [4]:
# Выделим признаки
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra']

In [5]:
# Выделим учебную выборку
features_train, features_test, target_train, target_test = train_test_split(features,target, test_size=.4, random_state=12345)

In [6]:
# Выделим тестовую и валидационную выборки (поделим тестовую из скрипта выше в соотношении 50/50)
features_valid, features_test, target_valid, target_test = train_test_split(features_test,target_test, test_size=0.5, random_state=12345) 

In [7]:
# Посмотрим на размеры выборок
# Обучающая
print(features_train.shape)
print(target_train.shape)

(1928, 4)
(1928,)


In [8]:
# Тестовая
print(features_test.shape)
print(target_test.shape)

(643, 4)
(643,)


In [9]:
# Валидационная
print(features_valid.shape)
print(target_valid.shape)

(643, 4)
(643,)


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Данные поделены, замечательно! Плюсом также является проверка размеров выборки, молодец
</div>

## Исследуйте модели

### Модель дерева решений

In [10]:
# Построим модель дерева решений, изменяя параметр глубины дерева от 1 до 10 
#с целью нахождения модели с максимальным показателем accuracy
best_model_tree = None
best_depth_tree = 0
best_accuracy_tree = 0
for depth in range(1,10):
    model_tree = DecisionTreeClassifier(random_state=12345, max_depth=depth).fit(features_train,target_train)
    predictions_tree = model_tree.predict(features_valid)
    score_tree = accuracy_score(target_valid,predictions_tree)
    if best_accuracy_tree<score_tree:
        best_model_tree = model_tree
        best_depth_tree = depth
        best_accuracy_tree = score_tree

In [11]:
# Выведем данные о лучшей модели
print(f'Лучший показатель depth: {best_depth_tree}, при нем показатель accuracy равен: {best_accuracy_tree}')

Лучший показатель depth: 3, при нем показатель accuracy равен: 0.7853810264385692


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Да, дерево хорошо обучилось)
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔:</b> 

Тут можно еще проверить гиперпараметр - атрибут разбиения (criterion), так как она осуществляет важную функцию - при обучении множества в узле, данный атрибут следит за тем, чтобы в этих множествах было меньше дисбаланса классов значений, к которым мы и предсказываем что-либо. Ну это так, на подумать, может в будущем ты захочешь его применить)
</div>

### Модель случайного леса

In [12]:
# Построим модель случайного леса, изменяя параметр глубины дерева и количество деревьев от 1 до 10 
#с целью нахождения модели с максимальным показателем accuracy
best_model_forest = None
best_n_forest = 0
best_depth_forest = 0
best_accuracy_forest = 0
for n in range(1,11):
    for depth in range(1,11):
        model_forest = RandomForestClassifier(random_state=12345,n_estimators=n,max_depth=depth).fit(features_train,target_train)
        predictions_forest = model_forest.predict(features_valid)
        score_forest = accuracy_score(target_valid,predictions_forest)
        if best_accuracy_forest < score_forest:
            best_model_forest = model_forest
            best_n_forest = n
            best_depth_forest = depth
            best_accuracy_forest = score_forest

In [13]:
# Выведем данные о лучшей модели
print(f'Лучший показатель depth: {best_depth_forest} с количество деревьев {best_n_forest}, при нем показатель accuracy равен: {best_accuracy_forest}')

Лучший показатель depth: 8 с количество деревьев 8, при нем показатель accuracy равен: 0.80248833592535


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Лес тоже сформирован и дал лучший результат (на данный момент), отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔:</b> 

Деревьев стоит выбрать побольше, так как именно от этого параметра зависит качество предсказания нашей модели, ведь лес - это множество моделей дерева, где учитывается каждое дерево, дающее свой результат. Попробуй выбрать значения побольше и шаг тоже добавь
    
А так можно еще взять гиперпараметр дерева - атрибут разбиения, о котором я выше говорил
</div>

### Модель логистической регрессии

In [14]:
model_regression = LogisticRegression(random_state = 12345).fit(features_train,target_train)
predictions_regression = model_regression.predict(features_valid)
score_regression = accuracy_score(target_valid,predictions_regression)
print(f'Показатель accuracy модели логистической регрессии равен {score_regression}')

Показатель accuracy модели логистической регрессии равен 0.7107309486780715


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Эх, логистическая регрессия показала не самый лучший результат. Возможно проблема в данных
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔:</b> 

К сожалению я не увидел здесь перебор гиперпараметров, а у логистической регрессии есть очень хорошие гиперпараметры, меняя которые можно немного поднять качество: сила регуляризации и веса признаков. Попробуй поизучать данные гиперпараметры и перебрать их) 
    
Если качество не улучшится, не стоит огорчаться, у нас данных очень мало
</div>

Таким образом, среди всех моделей наибольший показатель accuracy у модели случайного леса с 8 деревьями и показателем depth 8

## Проверьте модель на тестовой выборке

In [15]:
# Проверим лучшую модель на тестовой выборке
predictions_test = best_model_forest.predict(features_test)
accuracy_test = accuracy_score(target_test,predictions_test)
print(f'Показатель accuracy по тестовой выборке у наилучшей модели: {accuracy_test}')

Показатель accuracy по тестовой выборке у наилучшей модели: 0.7962674961119751


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Отлично, на тестовых данных результат хороший, задача выполнена, молодец)
</div>

## (бонус) Проверьте модели на адекватность

К сожалению, не успеваю выполнить это задание - хочу подтянуть знания по моделям, поэтому оставлю его пустым

<div class="alert alert-block alert-info">
<b>Совет🤔:</b> 
    
Ничего страшного, проверка на адекватность у вас будет потом, но если тебе интересно, то есть разные модели, построив которые можно реализовать проверку на адекватность, например, Dummy Models для более удобной и функциональной проверки на адекватность.
</div>

## Вывод

На основании выполненной работы опишем, какие шаги были выполнены и какие результаты получены:

    1) Был прочитан датафрейм, предобработка была совершена предварительно, поэтому изменения коснулись лишь приведения двух столбцов к другому типу.

    2) Была определена задача и на ее основе выбрани модели для рассмотрения: дерево решений, случайный лес и логистическая регрессия.

    3) Исходные данные были распределены на обучающую, тестовую и валидационные выборки в долях 60/20/20.

    3) Были построены несколько моделей, среди которых были выбраны лучшие на основании показателя accuracy. Наивысший показатель accuracy = 0,80 был получен у модели случайного леса с 8 деревьями и показателем depth = 8

    4) Наилучшая модель была применена к тестовой выборке - значение accuracy составило 0,796, что выше указанного порога 0,75

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Но вывод получился крайне классным - у тебя расписано все, что ты реализуешь во время выполнения проекта, кратко и сжато, мне нравится это. Мне понравилось то, что ты расписал все этапы проекта, которые были выполнены - так держать)
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75


<div style="border:solid Chocolate 2px; padding: 40px">

<h2><b>Общий вывод по проекту</b></h2>

Ты очень постарался, спасибо тебе за такую работу!)

В ходе проекта были выполнены все этапы проекта, главная цель достигнута. Проект выполнен грамотно, четко и понятно, ты отлично справился с ним).
    
- Есть советы по проекту, если хочешь и есть время - реализуй их. Если нет, то запомни на будущее)

Работа классная, проект принимаю. Удачи в следующих проектах!)